### PR TITLE
implemented cloud formation service

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards https://github.com/giantswarm/giantswarm/issues/2507.

### CODE REVIEW

File: service/awsconfig/v7/cloudformation/cloud_formation.go [hunk 1]
```diff
@@ -25,9 +26,51 @@ func New(config Config) (*CloudFormation, error) {
 	return c, nil
 }
 
-// TODO
-func (c *CloudFormation) DescribeOutputs(stackName string) ([]*cloudformation.Output, error) {
-	return nil, nil
+func (c *CloudFormation) DescribeOutputsAndStatus(stackName string) ([]*cloudformation.Output, string, error) {
```
Comment (xh3b4sd) [inline service/awsconfig/v7/cloudformation/cloud_formation.go]: In the first place I thought I can sort this in a different way but apparently I cannot. So we have to check the outputs and the status at the same time.

File: service/awsconfig/v7/key/key.go [hunk 2]
```diff
@@ -33,8 +33,16 @@ const (
...
+	MasterImageIDKey            = "MasterImageID"
+	MasterInstanceTypeKey       = "MasterInstanceType"
+	MasterCloudConfigVersionKey = "MasterCloudConfigVersion"
+	WorkerASGKey                = "WorkerASGName"
+	WorkerASGRef                = "workerAutoScalingGroup"
+	WorkerCountKey              = "WorkerCount"
+	WorkerImageIDKey            = "WorkerImageID"
+	WorkerInstanceTypeKey       = "WorkerInstanceType"
+	WorkerCloudConfigVersionKey = "WorkerCloudConfigVersion"
+	VersionBundleVersionKey     = "VersionBundleVersion"
```
Comment (xh3b4sd) [inline service/awsconfig/v7/key/key.go]: In order to be reused this needs to go out of the resource now.

File: service/awsconfig/v7/resource/cloudformation/current.go [hunk 3]
```diff
@@ -18,100 +18,75 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
 		return StackState{}, microerror.Mask(err)
 	}
 
-	// In order to compute the current state of the guest cluster's main stack we
-	// have to describe the CF stacks and lookup the right stack. We dispatch our
-	// custom StackState structure and enrich it with all information necessary to
-	// reconcile the cloudformation resource.
+	r.logger.LogCtx(ctx, "level", "debug", "message", "looking for the guest cluster cloud formation stack in the AWS API")
```
Comment (xh3b4sd) [inline service/awsconfig/v7/resource/cloudformation/current.go]: Note the log structure change.

File: service/awsconfig/v7/resource/cloudformation/current.go [hunk 4]
```diff
@@ -18,100 +18,75 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
...
-		}
-	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "found the guest cluster cloud formation stack in the AWS API")
 
 	// In case the current guest cluster is already being updated, we cancel the
 	// reconciliation until the current update is done in order to reduce
 	// unnecessary friction.
-	if status == cloudformation.ResourceStatusUpdateInProgress {
-		r.logger.LogCtx(ctx, "debug", fmt.Sprintf("main stack is in state '%s'", cloudformation.ResourceStatusUpdateInProgress))
+	if stackStatus == cloudformation.ResourceStatusUpdateInProgress {
```
Comment (xh3b4sd) [inline service/awsconfig/v7/resource/cloudformation/current.go]: This is a reason why we need to lookup the outputs and the status.

File: service/awsconfig/v7/resource/cloudformation/current.go [hunk 5]
```diff
@@ -133,7 +108,7 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
 			MasterInstanceType:       masterInstanceType,
 			MasterCloudConfigVersion: masterCloudConfigVersion,
 
-			WorkerCount:              workers,
+			WorkerCount:              workerCount,
```
Comment (xh3b4sd) [inline service/awsconfig/v7/resource/cloudformation/current.go]: Consistency.

File: service/awsconfig/v7/resource/cloudformation/error.go [hunk 6]
```diff
@@ -27,18 +25,6 @@ func IsNotFound(err error) bool {
...
-//
-// FIXME: The validation error returned by the CloudFormation API doesn't make
-// things easy to check, other than looking for the returned string. There's no
-// constant in aws go sdk for defining this string, it comes from the service.
-func IsStackNotFound(err error) bool {
-	if err == nil {
-		return false
-	}
-	return strings.Contains(microerror.Cause(err).Error(), "does not exist")
-}
```
Comment (xh3b4sd) [inline service/awsconfig/v7/resource/cloudformation/error.go]: Dead code.

File: service/awsconfig/v7/resource/cloudformation/resource.go [hunk 7]
```diff
@@ -88,16 +88,6 @@ func getCloudFormationTags(customObject v1alpha1.AWSConfig) []*awscloudformation
...
 
-func getStackOutputValue(outputs []*awscloudformation.Output, key string) (string, error) {
-	for _, o := range outputs {
-		if *o.OutputKey == key {
-			return *o.OutputValue, nil
-		}
-	}
-
-	return "", microerror.Maskf(notFoundError, "stack outpout value for key '%s'", key)
-}
```
Comment (xh3b4sd) [inline service/awsconfig/v7/resource/cloudformation/resource.go]: Dead code.

File: service/awsconfig/v7/resource/cloudformation/spec.go [hunk 8]
```diff
@@ -5,16 +5,6 @@ const (
...
 
-	masterImageIDOutputKey            = "MasterImageID"
-	masterInstanceTypeOutputKey       = "MasterInstanceType"
-	masterCloudConfigVersionOutputKey = "MasterCloudConfigVersion"
-	workersOutputKey                  = "WorkerCount"
-	workerImageIDOutputKey            = "WorkerImageID"
-	workerInstanceTypeOutputKey       = "WorkerInstanceType"
-	workerCloudConfigVersionOutputKey = "WorkerCloudConfigVersion"
-
-	versionBundleVersionOutputKey = "VersionBundleVersion"
```
Comment (xh3b4sd) [inline service/awsconfig/v7/resource/cloudformation/spec.go]: Dead code.

File: integration/util.go [hunk 9]
```diff
@@ -14,7 +14,7 @@ import (
 )
 
 const (
-	defaultTimeout = 500
+	defaultTimeout = 600
```
Comment (rossf7) [inline integration/util.go]: This is OK in the short term but we need to investigate if stack creation times are increasing by a large amount. Would be good to have an issue for that.

Comment (xh3b4sd) [reply integration/util.go]: Vol already created one. I added a comment. See https://github.com/giantswarm/giantswarm/issues/2717.

File: integration/framework.go [hunk 10]
```diff
@@ -350,17 +350,19 @@ func (f *framework) initGuestClientset() error {
...
 	}
 
 	if err := f.waitForAPIUp(); err != nil {
-		return microerror.Mask(err)
+		return microerror.Maskf(err, "waiting for API being up")
 	}
 
 	if err := f.WaitForNodesUp(minimumNodesReady); err != nil {
-		return microerror.Mask(err)
+		return microerror.Maskf(err, "waiting for nodes being up")
```
Comment (kopiczko) [inline integration/framework.go]: We discussed we should avoid adding annotations to returned errors. Is there something special about this case?

**Pull request merged**